### PR TITLE
refactor: move flavours into enum

### LINKED ISSUE #22: Rust: Flavours as enum variants
# Motivation
I want to pattern match on the current flavour.
```rust
// Fix latte colours
if matches!(flavour, Latte) {
	...
}
```

# Proposed solution
```rust
pub enum Flavour {
	Latte,
	...
}

impl Flavour {
	pub fn name(&self) -> &'static str { ... }

	pub fn palette(&self) -> Palette { ... }
	// OR
	pub fn rosewater(&self) -> Colour { ... }
	pub fn flamingo(&self) -> Colour { ... }
	pub fn pink(&self) -> Colour { ... }
	...
}
```

### PR DESCRIPTION
**Pull request by Gingeh**
fixes #22 

## Notes:
 - Palettes used in the `Flavour::palette` impl could be extracted into private items in separate files
 - This is a breaking change, so the major version number will need to be incremented before release

### CODE REVIEW

File: rust/src/lib.rs [hunk 1]
```diff
@@ -55,72 +55,64 @@
...
-pub use macchiato::MACCHIATO;
-
-mod mocha;
-pub use mocha::MOCHA;
-
 /// Returns an iterator over the four delicious Catppuccin flavours.
 pub fn flavours_iter() -> std::array::IntoIter<Flavour, 4> {
-    [LATTE, FRAPPE, MACCHIATO, MOCHA].into_iter()
+    [Flavour::Latte, Flavour::Frappe, Flavour::Macchiato, Flavour::Mocha].into_iter()
 }
```
Comment (backwardspy) [inline rust/src/lib.rs]: i think this should now just be an `iter` method on `Flavour`, what do you think?

**Review by backwardspy — COMMENTED** [review]
nice, i like it. i've commented on one function that i think can now be made more idiomatic thanks to these changes.

File: rust/src/lib.rs (hunk 1 above)
Comment (nyxkrage) [reply rust/src/lib.rs]: I agree with @backwardspy here, though the function should be called `into_iter()` since it's returning an `IntoIter` type, see [C-ITER-TY in the API guidelines](https://rust-lang.github.io/api-guidelines/naming.html#iterator-type-names-match-the-methods-that-produce-them-c-iter-ty)

**Review by nyxkrage — CHANGES_REQUESTED** [review]

File: rust/src/lib.rs (hunk 1 above)
Comment (backwardspy) [reply rust/src/lib.rs]: `into_iter` as suggested by @nyxkrage

File: rust/src/palette.rs [hunk 2]
```diff
@@ -0,0 +1,125 @@
...
+    pub surface1: Colour,
+    pub surface0: Colour,
+    pub base: Colour,
+    pub mantle: Colour,
+    pub crust: Colour,
+}
+
+impl Palette {
+    /// Returns an iterator over the colours in the flavour.
+    pub fn iter(&self) -> std::array::IntoIter<&Colour, 26> {
```
Comment (backwardspy) [inline rust/src/palette.rs]: `into_iter` here too

**Comment by nyxkrage** [comment]
This looks good to me now

**Review by backwardspy — APPROVED** [review]
lgtm!

File: rust/src/lib.rs [hunk 3]
```diff
@@ -55,72 +55,59 @@
...
+            assert_eq!(colours.next(), Some(&palette.subtext0));
+            assert_eq!(colours.next(), Some(&palette.overlay2));
+            assert_eq!(colours.next(), Some(&palette.overlay1));
+            assert_eq!(colours.next(), Some(&palette.overlay0));
+            assert_eq!(colours.next(), Some(&palette.surface2));
+            assert_eq!(colours.next(), Some(&palette.surface1));
+            assert_eq!(colours.next(), Some(&palette.surface0));
+            assert_eq!(colours.next(), Some(&palette.base));
+            assert_eq!(colours.next(), Some(&palette.mantle));
+            assert_eq!(colours.next(), Some(&palette.crust));
```
Comment (sgoudham) [inline rust/src/lib.rs:110]: I wonder if it's good to include a sad path test here ensuring that the next colour is actually of type `None`, I won't block the PR on this but maybe good for when refactoring next time

File: rust/src/palette.rs [hunk 4]
```diff
@@ -0,0 +1,125 @@
...
+    assert_eq!(next_colour(), flavour.subtext0);
+    assert_eq!(next_colour(), flavour.overlay2);
+    assert_eq!(next_colour(), flavour.overlay1);
+    assert_eq!(next_colour(), flavour.overlay0);
+    assert_eq!(next_colour(), flavour.surface2);
+    assert_eq!(next_colour(), flavour.surface1);
+    assert_eq!(next_colour(), flavour.surface0);
+    assert_eq!(next_colour(), flavour.base);
+    assert_eq!(next_colour(), flavour.mantle);
+    assert_eq!(next_colour(), flavour.crust);
```
Comment (sgoudham) [inline rust/src/palette.rs:124]: Same comment as above

**Review by sgoudham — APPROVED** [review]
Added some comments but will still approve the PR, thanks!
